### PR TITLE
Approve Redirect and Init a new Account

### PR DESCRIPTION
**Pull request by Lev-Stambler**
- Add approve redirect to sign and send transactions
- Add functionality to create a new account
- Flesh out multi chain to support accounts and signing
- Run npm format

### CODE REVIEW

File: libs/multi-chain/src/lib/near/account.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      near,
+      masterAccount,
+      urlAccountCreator,
+      keyStore,
+      this.initParams
+    );
+    return this.nearSingleton;
+  }
+
+  getAccountNameFromPubkey(pubkey: PublicKey, curve: CryptoCurves) {
```
Comment (Sladuca) [inline libs/multi-chain/src/lib/near/account.ts:72]: do you think we can make this static and make an overloaded version that isn't static? Static methods are easier to use for other purposes - for instance, if someone wanted to get the account id from a pubkey but didn't have a keyStore?

File: libs/multi-chain/src/lib/near/signer.ts [hunk 2]
```diff
@@ -6,12 +6,23 @@ import {
...
 import type { NearNetworkId } from '@baf-wallet/interfaces';
 import { sha256 } from 'js-sha256';
+import { Buffer } from 'buffer';
+import { formatKey } from '../utils';
+import { Action as NearNativeAction } from 'near-api-js/lib/transaction';
+import BN from 'bn.js';
 
 export interface NearSendTXOpts {
-  actions: transactions.Action[];
+  actions: NearAction[];
```
Comment (Sladuca) [inline libs/multi-chain/src/lib/near/signer.ts:25]: good

File: libs/multi-chain/src/lib/near/signer.ts [hunk 3]
```diff
@@ -42,8 +53,18 @@ export class NearSigner extends Signer<NearSendTXOpts> {
...
+        }
+        return transactions.transfer(
+          new BN((action.params as NearTransferParam).amount)
+        );
+      default:
+        throw `Action of type ${action.type} is unsupported`;
+    }
   }
 
   public async sendTX(opts: NearSendTXOpts) {
```
Comment (Sladuca) [inline libs/multi-chain/src/lib/near/signer.ts:70]: can we separate this into `createTx`, `signTx` and `sendTx` such that you can use one without the othesr, ideally making them default static but overload with a non-static method? (can you tell why I don't like OOP yet :P)

For instance, creating the transaction  `sendTx` in theory only requires a pubkey and networkId and I could imagine many cases where someone would want to create a tx without signing it (pretty much any Dapp). Another example is cases where someone would want to sign a tx without sending it directly to the near blockchain - for instance, maybe when using some layer 2.

**Review by Sladuca — CHANGES_REQUESTED** [review]
This be a thiccc PR! Left a bunch of comments, the bulk of them regarding making individual functionalities in multichain classes more decoupled so that you don't have to use the class and/or you can use only a small subset of the functionality using only the things that functionality requires.

File: apps/api/src/app/controllers/NearController.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import { createNearAccount } from '../service/near';
+import { CryptoCurves, PublicKey } from '@baf-wallet/interfaces';
+import { ChainUtil } from '@baf-wallet/multi-chain';
+import { ec } from 'elliptic';
+
+interface CreateNearAccountParams {
+  discordUser: string;
+  nonce: string;
+  signature: ec.Signature;
+  derivedEd25519Pubkey: PublicKey;
```
[the remark this answers is not in the document]
Comment (Lev-Stambler) [reply apps/api/src/app/controllers/NearController.ts:21]: Sure, that makes sense

File: apps/api/src/main.ts [hunk 5]
```diff
@@ -3,12 +3,27 @@
...
 
 const app = express();
 
+// TODO: put into constants once StoreKeys merged
+const defaultNearConfig = {
+  masterAccountId: 'levtester.testnet',
+  connectConfig: {
+    networkId: getNearNetworkId(Envs.DEV),
+    nodeUrl: 'https://rpc.testnet.near.org',
+    keyPath: '/home/lev/.near-credentials/testnet/levtester.testnet.json',
```
[the remark this answers is not in the document]
Comment (Lev-Stambler) [reply apps/api/src/main.ts:20]: Yeah, so env and env.production are for discriminating between envs. Constants (in the store keys branch) would hold stuff like the above. Store keys also takes care of .env stuff. I'll link this to a Github issue and we can resolve it once store keys is merged in

File: apps/frontend/.browserslistrc [hunk 6]
```diff
@@ -0,0 +1,3 @@
+last 1 version
+> 1%
+IE 10 # sorry
```
[the remark this answers is not in the document]
Comment (Lev-Stambler) [reply apps/frontend/.browserslistrc:3]: lmao, updated:
```
# Browsers that we support

defaults
not IE 11
maintained node versions
```

File: apps/frontend/src/components/base/Modal.svelte [hunk 7]
```diff
@@ -1,82 +1,85 @@
...
+        ...options,
+        props: {
+          ...props,
+          ...options.props,
+        },
+      });
+    };
+  }
 
-    import { setContext, createEventDispatcher } from 'svelte';
```
[the remark this answers is not in the document]
Comment (Lev-Stambler) [reply apps/frontend/src/components/base/Modal.svelte:21]: Yeah, got moved for some reason. Added back

File: libs/multi-chain/src/lib/near/account.ts [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+import { PublicKey } from '@baf-wallet/interfaces';
+import { formatKey } from '../utils';
+import { KeyPairEd25519 } from 'near-api-js/lib/utils';
+
+interface NearSingletonParams {
+  connectConfig: ConnectConfig;
+  masterAccountId: string;
+}
+
+export class NearAccountSingelton {
```
[the remark this answers is not in the document]
Comment (Lev-Stambler) [reply libs/multi-chain/src/lib/near/account.ts:22]: Yeah, thats better

File: libs/multi-chain/src/lib/near/account.ts (hunk 1 above)
Comment (Lev-Stambler) [reply libs/multi-chain/src/lib/near/account.ts:72]: There is a static version defined right below

File: libs/multi-chain/src/lib/near/signer.ts (hunk 3 above)
Comment (Lev-Stambler) [reply libs/multi-chain/src/lib/near/signer.ts:70]: I seperated out into different functions. I think adding a static overload for now might be over specing. Like, as long as its an easy thing to add later, we can add it when needed?

Comment (Sladuca) [reply libs/multi-chain/src/lib/near/signer.ts]: Sure, just keep it in mind and do it before it's too late.

**Review by Sladuca — APPROVED** [review]
Just keep modularity discussion from earlier in mind + the keyPath thing should get fixed by our keygen stub

**Pull request merged**